src/tests/tools/iotc_test_echoserver.cc

```diff
@@ -0,0 +1,164 @@
+/* Copyright 2018-2019 Google LLC
+ *
+ * This is part of the Google Cloud IoT Device SDK for Embedded C,
+ * it is licensed under the BSD 3-Clause license; you may not use this file
+ * except in compliance with the License.
+ *
+ * You may obtain a copy of the License at:
+ *  https://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "iotc_test_echoserver.h"
+
+#include <netinet/in.h>
+#include <string.h>
+#include <thread>
+
+namespace iotctest {
+
+EchoTestServer::EchoTestServer(uint16_t socket_type, uint16_t port,
+                               uint16_t protocol_type, const char* host)
+    : socket_type_(socket_type), test_port_(port),
+      protocol_type_(protocol_type), host_(host) {
+  CreateServer();
+}
+
+EchoTestServer::~EchoTestServer() {}
+
+void EchoTestServer::Run() {
+  if (server_thread_) {
+    EchoTestServer::Stop();
+  }
+  runnable_ = true;
+  if (socket_type_ == SOCK_STREAM) {
+    // TODO(b/127770330)
+    // server_thread_ =
+    //     std::make_unique<std::thread>(&EchoTestServer::RunTcpServer, this);
+    server_thread_ = std::unique_ptr<std::thread>(
+        new std::thread(&EchoTestServer::RunTcpServer, this));
+  } else if (socket_type_ == SOCK_DGRAM) {
+    // TODO(b/127770330)
```

Comment:
Got it. This would be out of scope of this PR, so leave this out, we can address it in another PR.